### PR TITLE
FINERACT-2090: Improve Loan Application Reject logic

### PR DESCRIPTION
**Pull request by ruchiD**
## Description
Refactor for Loan Reject logic.

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java [hunk 1]
```diff
@@ -794,19 +794,25 @@ public CommandProcessingResult rejectApplication(final Long loanId, final JsonCo
...
         final AppUser currentUser = getAppUserIfPresent();
 
-        this.loanApplicationTransitionApiJsonValidator.validateRejection(command.json());
+        // validate before loan assembling
+        loanApplicationTransitionValidator.validateRejection(command.json());
 
+        // retrieve/assemble loan
         final Loan loan = retrieveLoanBy(loanId);
 
+        // check client or group
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java]: Please move this into the `loanApplicationTransitionValidator.validateRejection`

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java [hunk 2]
```diff
@@ -794,19 +794,25 @@ public CommandProcessingResult rejectApplication(final Long loanId, final JsonCo
...
         checkClientOrGroupActive(loan);
 
+        // check for mandatory entities
         entityDatatableChecksWritePlatformService.runTheCheckForProduct(loanId, EntityTables.LOAN.getName(),
                 StatusEnum.REJECTED.getCode().longValue(), EntityTables.LOAN.getForeignKeyColumnNameOnDatatable(), loan.productId());
 
+        // validate loan rejection
+        loanApplicationTransitionValidator.validateRejection(loan, defaultLoanLifecycleStateMachine, command);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java]: Please merge together the loanApplicationTransitionValidator.validateRejection methods. I dont think we need 2.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java [hunk 3]
```diff
@@ -794,19 +794,25 @@ public CommandProcessingResult rejectApplication(final Long loanId, final JsonCo
...
+
         final Map<String, Object> changes = loan.loanApplicationRejection(currentUser, command, defaultLoanLifecycleStateMachine);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java]: Please extract out logic from loan.loanApplicationRejection method into this service.

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please kindly check my review!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java (hunk 2 above)
Comment (ruchiD) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java]: @adamsaghy  second validation is on loan while the first validation is on json. hence kept in two methods.

Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java]: We can do all the validations before executing the rejection. You can load the loan in the first validation method as well if needed. ;)

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java (hunk 1 above)
Comment (ruchiD) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java]: done

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java (hunk 2 above)
Comment (ruchiD) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java]: done

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java (hunk 3 above)
Comment (ruchiD) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java]: done

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**